hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/enrichers/ExitCallsEnricher.java

```diff
@@ -0,0 +1,116 @@
+package org.hypertrace.traceenricher.enrichment.enrichers;
+
+import com.google.common.collect.Maps;
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+import org.hypertrace.core.datamodel.ApiNodeEventEdge;
+import org.hypertrace.core.datamodel.AttributeValue;
+import org.hypertrace.core.datamodel.Event;
+import org.hypertrace.core.datamodel.StructuredTrace;
+import org.hypertrace.core.datamodel.shared.ApiNode;
+import org.hypertrace.core.datamodel.shared.trace.AttributeValueCreator;
+import org.hypertrace.traceenricher.enrichedspan.constants.EnrichedSpanConstants;
+import org.hypertrace.traceenricher.enrichedspan.constants.utils.EnrichedSpanUtils;
+import org.hypertrace.traceenricher.enrichment.AbstractTraceEnricher;
+import org.hypertrace.traceenricher.trace.util.ApiTraceGraph;
+
+public class ExitCallsEnricher extends AbstractTraceEnricher {
+
+  @Override
+  public void enrichTrace(StructuredTrace trace) {
+    Map<ByteBuffer, ApiExitCallInfo> exitCallInfo = computeApiExitCallCount(trace);
+    trace.getEventList().stream()
+        .filter(event -> exitCallInfo.containsKey(event.getEventId()))
+        .forEach(
+            event -> {
+              ApiExitCallInfo apiExitCallInfo = exitCallInfo.get(event.getEventId());
+              addEnrichedAttribute(
+                  event,
+                  EnrichedSpanConstants.API_EXIT_CALLS_ATTRIBUTE,
+                  AttributeValueCreator.create(apiExitCallInfo.getExitCallCount()));
+              addEnrichedAttribute(
+                  event,
+                  EnrichedSpanConstants.API_CALLEE_NAME_COUNT_ATTRIBUTE,
+                  AttributeValue.newBuilder()
+                      .setValueMap(apiExitCallInfo.getCalleeNameToCallCount())
+                      .build());
+            });
+  }
+
+  /**
+   * An api_node represents an api call in the trace It consists of api_entry span and multiple
+   * api_exit and internal spans
+   *
+   * <p>This method computes the count & breakup of exit calls for any given api (identified by
+   * api_entry_span) This count is a composition of 2 things 1. link between api_exit_span in
+   * api_node to api_entry_span (child of api_exit_span) in another api_node 2. exit calls to
+   * backend from api_exit_span in api_node
+   */
+  Map<ByteBuffer, ApiExitCallInfo> computeApiExitCallCount(StructuredTrace trace) {
+    ApiTraceGraph apiTraceGraph = new ApiTraceGraph(trace);
+    // event -> api exit call count for the corresponding api_node
+    Map<ByteBuffer, ApiExitCallInfo> eventToExitInfo = Maps.newHashMap();
```

Comment:
In order to reduce the amount of memory, we could have actually created a map of `apiId -> apiExitCallInfo`

Number of APIs will be less than than number of events by a considerable margin

Again, depends on you, if you want to refactor. And, that would mean the enricher depends on the EndpointEnricher

Reply:
But the exit information would have to sent per event itself in the StructuredTrace message?

Reply:
>But the exit information would have to sent per event itself

Yes. I was just thinking to save some memory for computation of exit call info. Nothing major. Not a blocker comment at all